loader/dd_library_loader.c

```diff
@@ -0,0 +1,355 @@
+}
+
+static void ddloader_unregister_module(void) {
+    zend_module_entry *ddtrace_injected = zend_hash_str_find_ptr(&module_registry, ZEND_STRL("ddtrace_injected"));
+    if (!ddtrace_injected) {
+        return;
+    }
+
+    // Set the MSHUTDOWN function to NULL to avoid it being called by zend_hash_str_del
+    ddtrace_injected->module_shutdown_func = NULL;
+    zend_hash_str_del(&module_registry, ZEND_STRL("ddtrace_injected"));
+}
+
+static PHP_MINIT_FUNCTION(ddtrace_injected) {
+    zend_module_entry *ddtrace = zend_hash_str_find_ptr(&module_registry, ZEND_STRL("ddtrace"));
+    if (ddtrace) {
+        LOG(INFO, "ddtrace is already loaded, unregister ddtrace_injected");
+        ddloader_unregister_module();
+
+        return SUCCESS;
+    }
+
+    LOG(INFO, "ddtrace is not loaded, checking the dependencies");
+
+    if (!ddloader_check_deps(orig_ddtrace_module_deps)) {
+        LOG(WARN, "ddtrace dependencies are not met, unregister ddtrace_injected");
+        ddloader_unregister_module();
+
+        return SUCCESS;
+    }
+
+    LOG(INFO, "Rename ddtrace_injected to ddtrace");
+
+    /**
+     * Rename the "key" of the module_registry to access ddtrace.
+     * Must be done at the bucket level to not change the order of the HashTable.
+     */
+    zend_string *old_name = ddloader_zend_string_init(php_api_no, ZEND_STRL("ddtrace_injected"), 1);
+    Bucket *bucket = (Bucket *)zend_hash_find(&module_registry, old_name);
+    ddloader_zend_string_release(php_api_no, old_name);
+
+    zend_string *new_name = ddloader_zend_string_init(php_api_no, ZEND_STRL("ddtrace"), 1);
+    ddloader_zend_hash_set_bucket_key(php_api_no, &module_registry, bucket, new_name);
+    ddloader_zend_string_release(php_api_no, new_name);
+
+    ddtrace = zend_hash_str_find_ptr(&module_registry, ZEND_STRL("ddtrace"));
+    if (!ddtrace) {
+        LOG(ERROR, "ddtrace not found. Something wrong happened");
+        return SUCCESS;
+    }
+
+    /* Restore name, MINIT, dependencies and functions of the module */
+    ddtrace->name = "ddtrace";
+    ddtrace->module_startup_func = origin_ddtrace_module_startup_func;
+    ddtrace->deps = orig_ddtrace_module_deps;
+    ddtrace->functions = orig_functions;
+    if (ddtrace->functions && zend_register_functions(NULL, ddtrace->functions, NULL, ddtrace->type) == FAILURE) {
+        LOG(ERROR, "Unable to register ddtrace's functions");
+        return SUCCESS;
+    }
+
+    return origin_ddtrace_module_startup_func(INIT_FUNC_ARGS_PASSTHRU);
+}
+
+static int ddloader_load_ddtrace(int php_api_no, char *module_build_id, bool is_zts, bool is_debug) {
+    char *ext_path = ddloader_find_ext_path("trace", "ddtrace", php_api_no, is_zts, is_debug);
+    if (!ext_path) {
+        LOG(ERROR, "Extension file not found");
+        return SUCCESS;
+    }
+
+    // The code below basically comes from the function "php_load_extension" in "ext/standard/dl.c",
+    // which does not allow loading an extension using a full path.
```

Comment:
That's news to me. php_load_extension reads:
```
	if (strchr(filename, '/') != NULL || strchr(filename, DEFAULT_SLASH) != NULL) {
		/* Passing modules with full path is not supported for dynamically loaded extensions */
		if (type == MODULE_TEMPORARY) {
			php_error_docref(NULL, E_WARNING, "Temporary module name should contain only filename");
			return FAILURE;
		}
		libpath = estrdup(filename);
```
Just not allowed for MODULE_TEMPORARY, but otherwise it works perfectly fine.

Reply:
```suggestion
    // The code below basically comes from the function "php_load_extension" in "ext/standard/dl.c",
    // but we need to rename the extension before passing it into the module_registry.
```